src/signTx.c

```diff
@@ -432,6 +488,150 @@ static void signTx_handleInitAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wi
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTx_handleAuxDataArbitraryHash_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+
+	UI_STEP(HANDLE_AUX_DATA_ARBITRARY_HASH_STEP_DISPLAY) {
+		ui_displayHexBufferScreen(
+		        "Auxiliary data hash",
+		        ctx->auxDataHash,
+		        SIZEOF(ctx->auxDataHash),
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_AUX_DATA_ARBITRARY_HASH_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceStage();
+	}
+	UI_STEP_END(HANDLE_AUX_DATA_ARBITRARY_HASH_STEP_INVALID);
+}
+
+
+enum {
+	HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_DISPLAY = 850,
+	HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_RESPOND,
+	HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_INVALID,
+};
+static void signTx_handleAuxDataCatalystRegistration_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTx_handleAuxDataCatalystRegistration_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+
+	UI_STEP(HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_DISPLAY) {
+		ui_displayPrompt(
+		        "Register Catalyst",
+		        "voting key?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		signTxCatalystRegistration_init();
+		ctx->stage = SIGN_STAGE_AUX_DATA_CATALYST_REGISTRATION_SUBMACHINE;
+	}
+	UI_STEP_END(HANDLE_AUX_DATA_CATALYST_REGISTRATION_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTx_handleAuxDataAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		TRACE_STACK_USAGE();
+		ASSERT(ctx->includeAuxData == true);
+
+		// delegate to state sub-machine for stake pool registration certificate data
+		if (signTxCatalystRegistration_isValidInstruction(p2)) {
+			TRACE();
+			CHECK_STAGE(SIGN_STAGE_AUX_DATA_CATALYST_REGISTRATION_SUBMACHINE);
+
+			TRACE_STACK_USAGE();
+
+			signTxCatalystRegistration_handleAPDU(p2, wireDataBuffer, wireDataSize);
+			return;
+		} else {
+			CHECK_STAGE(SIGN_STAGE_AUX_DATA);
+		}
+	}
+
+	{
+		explicit_bzero(&ctx->auxDataHash, SIZEOF(ctx->auxDataHash));
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+		VALIDATE(view_remainingSize(&view) >= 1, ERR_INVALID_DATA);
+		txAuxDataCtx->auxDataType = parse_u1be(&view);
+
+		switch (txAuxDataCtx->auxDataType) {
+		case AUX_DATA_TYPE_ARBITRARY_HASH:
+			// parse data
+			VALIDATE(view_remainingSize(&view) == AUX_DATA_HASH_LENGTH, ERR_INVALID_DATA);
+			STATIC_ASSERT(SIZEOF(ctx->auxDataHash) == AUX_DATA_HASH_LENGTH, "wrong metadata hash length");
+			view_memmove(ctx->auxDataHash, &view, AUX_DATA_HASH_LENGTH);
+			txAuxDataCtx->auxDataReceived = true;
+			break;
+		case AUX_DATA_TYPE_CATALYST_VOTING_KEY_REGISTRATION:
+			break;
+		default:
+			THROW(ERR_INVALID_DATA);
+		}
+
+		ASSERT(view_remainingSize(&view) == 0);
+	}
+
+
+	security_policy_t policy = policyForSignTxAuxData();
```

Comment:
I would consider adding auxDataType as an argument here, even if we do not use it. That way, we explicitly signalize our intent, otherwise it looks a bit like we might have forgot that there are two cases to distinguish.